promql/engine.go

```diff
@@ -697,117 +655,355 @@ func (ev *evaluator) eval(expr Expr) Value {
-		rhs := ev.evalOneOf(e.RHS, ValueTypeScalar, ValueTypeVector)
+	case *Call:
+		if e.Func.Name == "timestamp" {
+			// Matrix evaluation always returns the evaluation time,
+			// so this function needs special handling when given
+			// a vector selector.
+			vs, ok := e.Args[0].(*VectorSelector)
+			if ok {
+				return rangeWrapper(func(v []Value, enh *evalNodeHelper) Vector {
+					return e.Func.Call([]Value{ev.vectorSelector(vs, enh.ts)}, e.Args, enh)
+				})
+			}
+		}
+		// Check if the function has a matrix argument.
+		for matrixArgIndex, a := range e.Args {
+			sel, ok := a.(*MatrixSelector)
+			if ok {
+				inArgs := make([]Value, len(e.Args))
+				// Evaluate any non-matrix arguments.
+				otherArgs := make([]Matrix, len(e.Args))
+				otherArgsIn := make([]Vector, len(e.Args))
+				for i, e := range e.Args {
+					if i != matrixArgIndex {
+						otherArgs[i] = ev.eval(e).(Matrix)
+						otherArgsIn[i] = Vector{Sample{}}
+						inArgs[i] = otherArgsIn[i]
+					}
+				}
 
-		switch lt, rt := lhs.Type(), rhs.Type(); {
-		case lt == ValueTypeScalar && rt == ValueTypeScalar:
-			return Scalar{
-				V: scalarBinop(e.Op, lhs.(Scalar).V, rhs.(Scalar).V),
-				T: ev.Timestamp,
+				mat := make(Matrix, 0, len(sel.series)) // Output matrix.
+				offset := durationMilliseconds(sel.Offset)
+				// Reuse objects across steps to save memory allocations.
+				points := getPointSlice(16)
+				inMatrix := make(Matrix, 1)
+				inArgs[matrixArgIndex] = inMatrix
+				enh := &evalNodeHelper{out: make(Vector, 0, 1)}
+				// Process all the calls for one time series at a time.
+				var it *storage.BufferedSeriesIterator
+				for i, s := range sel.series {
+					if it == nil {
+						it = storage.NewBuffer(s.Iterator(), durationMilliseconds(sel.Range))
+					} else {
+						it.Reset(s.Iterator())
+					}
+					ss := Series{
+						// For all range vector functions, the only change to the
+						// output labels is dropping the metric name so just do
+						// it once here.
+						Metric: dropMetricName(sel.series[i].Labels()),
+						Points: getPointSlice(numSteps),
+					}
+					inMatrix[0].Metric = sel.series[i].Labels()
+					for ts, step := ev.Timestamp, -1; ts <= ev.EndTimestamp; ts += ev.Interval {
+						step++
+						// Set the non-matrix arguments.
+						// They are scalar, so it is safe to use the step number
+						// when looking up the argument, as there will be no gaps.
+						for j := range e.Args {
+							if j != matrixArgIndex {
+								otherArgsIn[j][0].V = otherArgs[j][0].Points[step].V
+							}
+						}
+						maxt := ts - offset
+						mint := maxt - durationMilliseconds(sel.Range)
+						// Evaluate the matrix selector for this series for this step.
+						points = ev.matrixIterSlice(it, maxt, mint, points[:0])
+						if len(points) == 0 {
+							continue
+						}
+						inMatrix[0].Points = points
+						enh.ts = ts
+						// Make the function call.
+						outVec := e.Func.Call(inArgs, e.Args, enh)
+						enh.out = outVec[:0]
```

Comment:
Same as above, would be more readable as

    enh.out = enh.out[:0]

and done before the function call.